File: runtime/orchestrator.py
"""This module defines the Orchestrator agent and its planning capabilities.

The Orchestrator is a specialized agent responsible for generating, managing,
and executing the workflow plan based on a high-level goal. This is being
refactored to move workflow execution to the System class.
"""

import os
import json
from typing import List, Dict, Optional, Tuple
import logging

from pydantic import BaseModel, Field, ValidationError

from runtime.agent import Agent
from runtime.core import Session
from runtime.util import read_file

logger = logging.getLogger(__name__)

from runtime.custom_types import Plan, File, AgentProfile


class PlanningPrompt(BaseModel):
    agents: list[AgentProfile] = Field(..., default_factory=list)
    files: list[File] = Field(..., default_factory=list)

    def make(self, goal: str) -> str:
        # Construct a description of the available agents and their roles/goals
        team_description = "\n".join(
            f"- Name: '{agent.name}'\n"
            f"  Role: `{agent.role}`\n"
            f"  Goal: \"{agent.goal}\""
            for agent in self.agents
        )

        # Load the general planning prompt template
        t20_root = os.path.abspath(os.path.join(os.path.dirname(__file__), '..')) # Not used
        general_prompt_template = read_file(os.path.join(t20_root, "prompts", "general_planning.txt")).strip()

        # Format the planning prompt parts
        planning_prompt_parts = [general_prompt_template.format(
            high_level_goal=goal,
            team_description=team_description
        )]

        # Include file contents if provided
        file_section = []
        for file in self.files:
            file_section.append(f"⫻context/file:{file.path}\n{file.content}\n\n\n")
        planning_prompt_parts.append("\n".join(file_section))

        logger.debug("Planning prompt parts for LLM: %s", planning_prompt_parts)

        # Combine prompt parts for the LLM call
        return "\n\n\n".join(planning_prompt_parts)


class Orchestrator(Agent):
    """An agent responsible for creating and managing a plan for multi-agent workflows."""
    team: Dict[str,Agent] = {}

    def generate_plan(self, session: Session, high_level_goal: str, files: List[File] = []) -> Optional[Plan]:
        """
        Invokes the language model to get a structured plan for the given high-level goal.
        The plan includes a sequence of tasks and the roles responsible for them.

        Args:
            session (Session): The current session object.
            high_level_goal (str): The high-level goal for which to generate a plan.
            file_contents (List[Tuple[str, str]]): A list of file paths and their contents to provide context to the LLM.

        Returns:
            Optional[Plan]: The generated plan as a Pydantic object, or None if an error occurs.
        """
        logger.info(f"Orchestrator {self.profile.name} is generating a plan for: '{high_level_goal}'")
        if not self.llm or not self.team:
            logger.error("Orchestrator client or team not initialized. Cannot generate plan.")
            return None

        planning_prompt = f"""
You are an expert orchestrator. Your goal is to break down a high-level goal into a detailed, actionable plan for a team of specialized agents.
The plan should be a sequence of tasks, each assigned to a specific agent role.

Here's the high-level goal: '{high_level_goal}'

Here are the available agents and their roles:
{json.dumps([{"name": agent.profile.name, "role": agent.profile.role, "goal": agent.profile.goal} for agent in self.team.values()], indent=4)}

Here are the files provided by the user:
{json.dumps([file.model_dump() for file in files], indent=4)}

Your plan should adhere to the following structure (Pydantic Plan model):
{Plan.model_json_schema(by_alias=False, mode='serialization')}

Key considerations for your plan:
1.  **Break down the goal**: Decompose the `high_level_goal` into smaller, manageable `tasks`.
2.  **Assign roles**: Each `task` must be assigned to an `agent` by its `name` and/or `role`. Ensure the assigned agent's goal aligns with the task's requirements.
3.  **Dependencies**: Define `requires` for each task, indicating which previous tasks must be completed before the current one can start. Use the `id` of the preceding tasks. The first task should have `requires: []`.
4.  **Comprehensive**: Ensure all aspects of the `high_level_goal` are covered by the tasks, hence `reasoning`.
5.  **Logical flow**: The sequence of tasks should be logical and progressive towards achieving the overall goal.
6.  **Output**: Your response MUST be a JSON object that strictly conforms to the `Plan` schema.
7.  **Team Updates (Optional)**: If necessary, you can include `team` updates in the plan, such as new `system_prompts` for agents or general `notes` about team coordination.
    -   If you update a `system_prompt`, ensure the `agent` name matches an existing agent exactly.
    -   The `system_prompt` should be the full, new system prompt, not just a diff.
"""

        planning_prompt = PlanningPrompt(agents=[t.profile for t in self.team.values()], files=files).make(high_level_goal)

        logger.info(f"Orchestrator {self.profile.name} Planning Prompt:\n{planning_prompt}")

        session.add_artifact("planning_prompt.txt", planning_prompt)

#        response = self.llm.generate_content(
#            model_name=self.model,
#            contents=planning_prompt,
#            system_instruction=self.system_instruction,
#            temperature=0.0
#        )

#        session.add_artifact("planning_response.txt", response)

        try:
            response = self.llm.generate_content(
                model_name=self.model,
                contents=planning_prompt,
                system_instruction=self.system_instruction,
                temperature=0.0,
                response_mime_type='application/json',
                response_schema=Plan
            )
            if isinstance(response, Plan):
                plan = response
            else:
                plan = Plan.model_validate_json(response or '{}')
        except (ValidationError, json.JSONDecodeError) as e:
            logger.exception(f"Error generating or validating plan for {self.profile.name}: {e}")
            return None
        except Exception as e:
            logger.exception(f"An unexpected error occurred during plan generation for {self.profile.name}: {e}")
            return None

        return plan

File: new-logs/session_69ca6e89-582d-40f0-aef4-a6ad2771df81/src/orchestrator/Orchestrator.ts
import os from 'os';
import path from 'path';
import fs from 'fs';
import yaml from 'js-yaml';
import { glob } from 'glob';

import { Agent, AgentSpec, PromptSpec } from '../agent/Agent';
import { Session } from '../core/Session';
import { ExecutionContext } from '../core/ExecutionContext';
import { Plan, File, Task, Role, Artifact } from '../custom_types';
import { LLM } from '../llm/LLM';
import { Logger } from '../logger/Logger';
import { read_file } from '../util/Util'; // Utility to read files

const logger = Logger.getInstance();

export class Orchestrator extends Agent {
  public team: Record<string, Agent> = {}; // Agents delegated to by this orchestrator

  constructor(name: string, role: string, goal: string, model: string, systemPrompt: string) {
    super(name, role, goal, model, systemPrompt);
    logger.info(`Orchestrator instance created: ${this.name} (Role: ${this.role}, Goal: ${this.goal}, Model: ${this.model})`);
  }

  /**
   * Generates a plan for the given high-level goal, considering available agents and context.
   * @param session The current session.
   * @param highLevelGoal The overall objective of the workflow.
   * @param files Initial files provided to the system.
   * @returns A Plan object or null if generation fails.
   */
  public async generatePlan(session: Session, highLevelGoal: string, files: File[] = []): Promise<Plan | null> {
    logger.info(`Orchestrator ${this.name} is generating a plan for: '${highLevelGoal}'`);

    if (!this.llm) {
      logger.error('LLM client is not initialized for the orchestrator.');
      return null;
    }

    // Construct a description of the available agents and their roles/goals for the prompt
    const teamDescription = Object.values(this.team).map(agent =>
      `    - Name: '${agent.name}'\n      Role:\n        title: '${agent.role}'\n        purpose: '${agent.goal}'`
    ).join('\n');

    // Load the general planning prompt template
    // NOTE: This path might need adjustment based on where prompts are stored relative to the executing script.
    const promptTemplatePath = path.resolve(__dirname, '..', 'prompts', 'general_planning.txt');
    const generalPromptTemplate = read_file(promptTemplatePath);

    if (!generalPromptTemplate || generalPromptTemplate.startsWith('Error:')) {
      logger.error(`Failed to load planning prompt template from ${promptTemplatePath}.`);
      return null;
    }

    // Format the prompt with goal, team description, and initial files
    let planningPrompt = generalPromptTemplate.replace('{high_level_goal}', highLevelGoal).replace('{team_description}', teamDescription);

    if (files.length > 0) {
      const fileSection = files.map(file =>
        `--- File: '${file.path}'\n${file.content}`
      ).join('\n\n');
      planningPrompt += `\n\nPlease consider the following input files:\n${fileSection}\n\n`;
    }

    logger.debug(`Planning prompt for LLM:\n${planningPrompt}`);
    await session.recordArtifact('planning_prompt.txt', planningPrompt);

    try {
      const response = await this.llm.generateContent(
        this.model, // Use the orchestrator's model
        planningPrompt,
        this.systemPrompt, // Use orchestrator's system prompt
        0.1, // Lower temperature for more deterministic planning
        'application/json', // Expect JSON output
        Plan // Specify the expected response schema
      );

      let plan: Plan;
      if (response instanceof Plan) {
        plan = response;
      } else {
        // Attempt to parse if not already a Plan instance (e.g., if response is JSON string)
        plan = Plan.parse(response); // Use Pydantic's parse method
      }

      logger.info('Successfully generated plan.');
      return plan;

    } catch (error) {
      logger.error(`Error generating plan: ${(error as Error).message}`, error);
      // Optionally, record the error or a partial plan if available
      await session.recordArtifact('plan_generation_error.txt', (error as Error).message);
      return null;
    }
  }

  /**
   * Executes a task within the context of the current execution.
   * This method is typically overridden by specific agent types.
   * For Orchestrator, it might involve delegating or performing planning actions.
   * @param context The current execution context.
   * @returns The result of the task execution.
   */
  public async executeTask(context: ExecutionContext): Promise<string | null> {
    // If the current step requires generating a plan, call generatePlan
    if (context.currentStep().description.toLowerCase().includes('generate plan')) {
      const plan = await this.generatePlan(context.session, context.plan.highLevelGoal, context.getInitialFiles());
      if (plan) {
        // Record the generated plan as part of the context's output for this step
        const artifactFile = new File({path: 'generated_plan.json', content: JSON.stringify(plan, null, 2)});
        return JSON.stringify(new AgentOutput({ output: 'Plan generated successfully.', artifact: { task: context.currentStep().id, files: [artifactFile] }, reasoning: 'Orchestrator completed plan generation task.' }));
      } else {
        return JSON.stringify(new AgentOutput({ output: 'Failed to generate plan.', reasoning: 'Orchestrator failed during plan generation.' }));
      }
    }

    // Default execution: delegate to base Agent or handle specific orchestrator logic
    logger.warn(`Orchestrator ${this.name} is executing a generic task. Default behavior is to delegate or plan.`);
    // Placeholder response if no specific logic is matched
    return JSON.stringify(new AgentOutput({ output: `Orchestrator ${this.name} placeholder task execution.`, reasoning: 'Default orchestrator task handling.' }));
  }
}
